File: LeetCode/23. Merge k Sorted Lists.cpp
// Author : @AdarshGaur

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int,ListNode*>, vector<pair<int,ListNode*>>, greater<>> pq;
        for(auto &list : lists){
            if(list != NULL){
                pq.push({list->val, list});
            }
        }
        
        ListNode* dummy = new ListNode(0);
        ListNode* head = dummy;
        while(!pq.empty()){
            auto [v, p] = pq.top();
            pq.pop();
            
            head->next = p;
            p = p->next;
            if(p != NULL)
                pq.push({p->val, p});
            head = head->next;
        }
        return dummy->next;
    }
};

File: LeetCode/23. Merge k Sorted Lists.py
#Author: @AdarshGaur

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

from queue import PriorityQueue

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        pq = PriorityQueue()
        for i in range(len(lists)):
            if lists[i]:
                pq.put((lists[i].val, i))
        
        dummy = ListNode(0)
        head = dummy
        while not pq.empty():
            val, i = pq.get()
            head.next = ListNode(val)
            head = head.next
            if lists[i].next:
                lists[i] = lists[i].next
                pq.put((lists[i].val, i))
        
        return dummy.next
        
